### DISCUSSION TITLE
Sorting Wiki

### DISCUSSION THREAD

**Discussion by ClamsyHub** [Leantime/leantime#2370] [Ideas]
Hi,

Question regarding the WIKI. Is it possible to do an individual sorting here?

The wiki is created "on the fly" and it can happen that a new entry should be higher up in the list. However, the entries are sorted from top to bottom according to the “creation date” or “ID”.

So e.g.
ID 1
ID2
ID3
ID4
ID5

But I would like to enter a new Wiki entry ID 6 between ID 2 and ID 3, but I can't manage that. In order for me to be able to do this currently, I would have to delete ID 3, ID 4, ID 5 - create the new entry ID 6, then create ID 3 (which would then be ID 7), recreate ID 4 (which would then be ID 8) and then ID 5 (which then becomes ID 9).

The bigger the wiki gets, the more complex it becomes. Can you please activate manual sorting here? Move Drap/Drop, or rearrange using the up/down arrow?

That would be great if this could be done in one of the next versions.

**Comment by ObySamKenoby** [comment]
I would like to upvote this.
Having the pages in random order (I know it is not random, anyway...) is really a problem.
For a bit of relief just modifying the query that fetches the headlines (app/domain/wiki/repositories/wiki.php, line 134 in version 3.4.9) from 

`ORDER BY parent DESC, sortindex`

to

`ORDER BY title`

makes the tree on the left ordered alphabetically and you can "move" the pages putting a number before them (10, 20...)

![image](https://github.com/user-attachments/assets/98495ad7-997b-436d-ba73-355c3e010de7)

The problem is still present in the dropdown, where the absence of grouping puts everything on the same level

![image](https://github.com/user-attachments/assets/ef45c7a2-0065-46ac-b174-75ccaa1be13c)

Still I find this preferable than having everything shuffled together.

I don't know PHP, anyway I'll try something on the WE to make the dropdown show something similar to a tree... Hope this can help.

**NB: I didn't see any countereffect modifying the query, but I use just a fraction of LT  potential. 
If you find that the change creates problems pls let me know.**

From what I can say you are already working on manual sorting (the sortindex column make me thinking this...)

Thanks for your work.

**Reply by marcelfolaron** [reply]
Thanks, that is actually a good point and fairly easy to accomplish.

**Reply by ObySamKenoby** [reply]
> Thanks, that is actually a good point and fairly easy to accomplish.

You are welcome

**Comment by ObySamKenoby** [comment]
**I updated the code to prevent circular references.**

I managed to have a result....
As I said I don't know PHP (I'm a c# programmer),  so the code can certainly be improved by much.

The lines are referenced to version 3.4.9

What is needed is to modify the query on **app/domain/wiki/repositories/wiki.php**, line 134 from
`ORDER BY parent DESC, sortindex DESC`
to
`ORDER BY parent, title`
This will 'fix' the wiki homepage:

![image](https://github.com/user-attachments/assets/91edc20a-4e36-41e7-b8a7-aa5b4f50f74d)

Then in 
apps\leantime\app\Domain\Wiki\Templates\articleDialog.tpl you should change the first PHP block (the on on top of the page) to

```php
<?php
foreach ($__data as $var => $val) {
    $$var = $val; // necessary for blade refactor
}
$currentArticle = $tpl->get('article');

/*
Original line:

$wikiHeadlines = $tpl->get('wikiHeadlines');

The following block orders and formats the headlines in the combobox.
*/

// --------------------------

$wikiHL = $tpl->get('wikiHeadlines');

$wikiHeadlines = [];

// Adds to $wikiHeadlines all the options and suboptions from $wikiHL for articles with parent $parentId.
// The method is called recursively.
// $id: current page Id.
// $parentId: the id of parent article.
// $wikiHeadline: the output, ordered array. will be modified inside the method.
// $wikiHL: the array returned from Service.getAllWikiHeadlines. 
//          Even if it is passed by reference for performance the method will not modify it.
// $indent is the string to put before the title, any level will add a space.
function createTree($id, $parentId, &$wikiHeadlines, &$wikiHL, $indent) {
	// Finds the first article 
	$articles = array_filter($wikiHL, function($v) use($parentId) { return $v->parent == $parentId; });
	if (count($articles) > 0) {
		usort ($articles, function($a1, $a2) { return $a1->title > $a2->title; });
		if ($parentId != null) {
			$indent = $indent . "-";
		}
		foreach ($articles as $article) {
			// This check prevents circular references by hiding the current page and its childs from list.
			if ($article->id != $id) {				
				$art = $article;
				$art->title = $indent . $article->title;
				$wikiHeadlines[] = $art;
				createTree($id, $article->id, $wikiHeadlines, $wikiHL, $indent);
			}
		}
	}
	//var_dump($articles);
	
}

// --------------------------

// The following is the second original php block
if (! isset($_GET['closeModal'])) {
    echo $tpl->displayNotification();
}

$id = '';
if (isset($currentArticle->id)) {
    $id = $currentArticle->id;
}

// Populates the options tree
createTree($id, null, $wikiHeadlines, $wikiHL, "");

?>
```

Then the combobox will look this way

![image](https://github.com/user-attachments/assets/074e016b-b79d-4015-b828-7ff661fb0f9b)

The numbers are there to order and left from previous tests, on the sublevels there is no need for the parent number.
Hope this can help.

**Reply by marcelfolaron** [reply]
Awesome, do you want to create a PR for this so you get the bragging rights, fame and credit :) ?

**Reply by ObySamKenoby** [reply]
> Awesome, do you want to create a PR for this so you get the bragging rights, fame and credit :) ?

I fixed the circular references problem.
I've got no clue on how to create a PR, but I'd love to. I'm a pro programmer but also a solo programmer, so I use only minimal part of Git :D.
If there is an hint on how to do, or if you want to do it yourself it's ok. I just wanted to contribute and my code is less than perfect...

**Reply by marcelfolaron** [reply]
If you want to try:
1. Click the "fork" button on the top right hand side of the leantime repository. This will create a copy of the repo in under your account.
2. Clone your repo (or make the changes directly using github UI)
3. Create a brach
4. Then going back to the leantime repository head to https://github.com/Leantime/leantime/pulls and click the "New Pull Request button
5. It should show you your own branch and how it would be merged into Leantime. Then click create. 

If that doesn't work (or is too much effort for now) I'll create the PR and add you as collaborator.

**Reply by ObySamKenoby** [reply]
Di the Pull request, thank you for your time.

Have a nice day.
